docs/subsystems/billing.md

```diff
@@ -41,3 +41,36 @@ Notes:
 * 4242424242424242 is Stripe's test credit card, also useful for manually
   testing. You can put anything in the address fields, any future expiry
   date, and anything for the CVV code.
+
+## RealmAuditLogProcessor
+
+The general strategy here is that billing-relevant events get written to
+RealmAuditLog with `requires_billing_update = True`, and then a queue worker
+goes through, reads RealmAuditLog row by row, and makes the appropriate
+updates in Stripe (in order), keeping track of its state in
+`RealmAuditLogProcessor`. An invariant is that it cannot be important when
+exactly the queue worker gets around to making the update in Stripe, as long
+as the updates for each customer (realm) are made in `RealmAuditLog.id` order.
+
+Almost all the complexity in the code is due to error handling. We
+distinguish two kinds of errors: "Customer agnostic" errors like network
+errors, Stripe being down, misconfigured API keys, etc. in one bucket and
+"Customer specific" errors like their card being declined in another.
+
+For customer agnostic errors we currently just pause the whole RealmAuditLog
+processing and restart after a bit.
+
+To handle customer specific errors we use the following strategy.  There is
+a global RealmAuditLogProcessor (with `realm=None`) that processes
+RealmAuditLog entries for every customer (realm). If it runs into a
+customer-specific issue on some entry, it gives up on that entry and
+(temporarily) all future entries of that realm, and spins off another
+RealmAuditLogProcessor (with non-null `realm`) that marks that realm as
+needing manual attention. When whatever issue has been corrected, the
+realm-specific RealmAuditLogProcessor completes any realm-specific
+RealmAuditLog entries, and then deletes itself.
```

Comment:
There isn't code to delete itself yet, right?

Reply:
correct